src/devices/bus/pce_ctrl/joypad6.cpp

```diff
@@ -0,0 +1,295 @@
+// license:BSD-3-Clause
+// copyright-holders:cam900
+/**********************************************************************
+
+    NEC PC Engine/TurboGrafx-16 6 button joypad emulation
+
+    Based on SMS controller port emulation (devices\bus\sms_ctrl\*.*)
+    by Fabio Priuli,
+    PC engine emulation (mame\*\pce.*)
+    by Charles MacDonald, Wilbert Pol, Angelo Salese
+
+    There's 2 officially licensed 6 button joypad models:
+
+    NEC Avenue Pad 6 (NAPD-1002)
+    - Supports autofire for Button I, II, Run (slow motion)
+
+    NEC Arcade Pad 6 (PCE-TP1)
+    - Supports autofire for Button I-VI, Bundled in PC Engine Duo-RX
+
+    TODO:
+    - Needs verifications for Part numbers
+    - Verify DTC114Y function for Slow motion in Avenue Pad 6
+
+**********************************************************************/
+
+#include "emu.h"
+#include "joypad6.h"
+
+
+
+//**************************************************************************
+//  DEVICE DEFINITIONS
+//**************************************************************************
+
+DEFINE_DEVICE_TYPE(PCE_AVENUE_PAD_6, pce_avenue_pad_6_device, "pce_avenue_pad_6", "NEC Avenue Pad 6")
+DEFINE_DEVICE_TYPE(PCE_ARCADE_PAD_6, pce_arcade_pad_6_device, "pce_arcade_pad_6", "NEC Arcade Pad 6")
+
+
+static INPUT_PORTS_START( pce_joypad6 )
+	PORT_START("BUTTONS_0")
+	PORT_BIT( 0x01, IP_ACTIVE_LOW, IPT_BUTTON3 ) PORT_NAME("Button I") // Rightmost in bottom row
+	PORT_BIT( 0x02, IP_ACTIVE_LOW, IPT_BUTTON2 ) PORT_NAME("Button II")
+	PORT_BIT( 0x04, IP_ACTIVE_LOW, IPT_SELECT  ) PORT_NAME("Select")
+	PORT_BIT( 0x08, IP_ACTIVE_LOW, IPT_START   ) PORT_NAME("Run")
+
+	PORT_START("DIRECTION")
+	PORT_BIT( 0x01, IP_ACTIVE_LOW, IPT_JOYSTICK_UP )    PORT_8WAY
+	PORT_BIT( 0x02, IP_ACTIVE_LOW, IPT_JOYSTICK_RIGHT ) PORT_8WAY
+	PORT_BIT( 0x04, IP_ACTIVE_LOW, IPT_JOYSTICK_DOWN )  PORT_8WAY
+	PORT_BIT( 0x08, IP_ACTIVE_LOW, IPT_JOYSTICK_LEFT )  PORT_8WAY
+
+	PORT_START("BUTTONS_1")
+	PORT_BIT( 0x01, IP_ACTIVE_LOW, IPT_BUTTON1 ) PORT_NAME("Button III") // Leftmost in bottom row
+	PORT_BIT( 0x02, IP_ACTIVE_LOW, IPT_BUTTON4 ) PORT_NAME("Button IV")
+	PORT_BIT( 0x04, IP_ACTIVE_LOW, IPT_BUTTON5 ) PORT_NAME("Button V")
+	PORT_BIT( 0x08, IP_ACTIVE_LOW, IPT_BUTTON6 ) PORT_NAME("Button VI")
```

Comment:
I’m still not convinced trying to get left-to-right buttons is the right approach.  In the current state, there’s no a pad with no button 1 at all, which is likely to confuse people as much as anything.  I know the Avenue Pad 6 is weird, having the top row of buttons ascending in the opposite order to the bottom row, but is this really better than just mapping button I to button 1, and so on?

Reply:
Looking at this code fragment, my eye is drawn to the assignment of IPT_BUTTON1 (which I assume is related to key/joypad bindings) against "Button III".  This would not make sense, as very few games can actually use a 6-button joypad (1 HuCard and 19 CDROM, of a library of roughly 400 and 475),

Assigning IPT_BUTTON1 to one of "Button I" or "Button II" makes more sense, and either one might be defensible depending on the argument, but keep in mind that the joypad layout for TGFX-16 is II/I in the same way that SNES is B/A (for consistency).